### PR TITLE
Using DynamicOneDimensionalArray in Arraystack

### LINKED ISSUE #67: Use DynamicOneDimensionalArray in ArrayStack
#### Description of the problem

Currently we use fixed size arrays in `ArrayStack` which is nothing but wastage of memory. To reduce that problem, `DynamicOneDimensionalArray`should be used inside the methods of `ArrayStack`.

The way of using is left to the one who will fix the issue.

### PR DESCRIPTION
**Pull request by Saptashrungi**
Fixes #67.

### CODE REVIEW

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+from pydatastructs.linear_data_structures import DynamicOneDimensionalArray
 from pydatastructs.linear_data_structures import OneDimensionalArray
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: This line should be deleted. We don't need `OneDynamicDimensionalArray` anymore.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 2]
```diff
@@ -81,47 +82,35 @@ def peek(self):
...
 
     @property
     def is_empty(self):
-        return self.top == 0
+        return self._last_pos_filled == 0
 
     @property
     def peek(self):
-        return self.items[self.top - 1]
+        return self.items[self._last_pos_filled-1]
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]:
```python
return self.items[self._last_pos_filled]
# No need to subtract `-1`, as `_last_pos_filled` points to the topmost element already, earlier top used to point one position above the topmost element.
```

**Comment by czgdp1807** [comment]
The API is breaking by your changes, that's why the tests are failing. We need to find out a way to keep the API intact while using `DynamicOneDimensionalArray`.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 3]
```diff
@@ -81,47 +82,35 @@ def peek(self):
...
+        obj.size, obj.items, obj.dtype = \
+            size, items, items._dtype
         return obj
 
     def push(self, x):
-        if self.top == self.maxsize:
-            raise ValueError("Stack is full.")
-        self.items[self.top] = self.dtype(x)
-        self.top += 1
+        self.items.append(dtype.x)
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: Just do `self.items.append(x)`. Casting should be done by the `DODA`, not a concern of stack.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 4]
```diff
@@ -81,47 +82,35 @@ def peek(self):
...
-            raise ValueError("maxsize is missing.")
-        if not _check_type(top, int):
-            raise TypeError("top is not of type int.")
+    def __new__(cls, size=None, items=None, dtype=int):
+        if not _check_type(size, int):
+            raise ValueError("size is missing.")
         if items is None:
-            items = OneDimensionalArray(dtype, maxsize)
-        if not _check_type(items, OneDimensionalArray):
+            items = DynamicOneDimensionalArray(dtype, size)
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: Do we even need `size` too? We can remove it safely. APIs will break which will require changes in other modules too.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 5]
```diff
@@ -81,47 +82,35 @@ def peek(self):
...
-    __slots__ = ['maxsize', 'top', 'items', 'dtype']
-
-    def __new__(cls, maxsize=None, top=0, items=None, dtype=int):
-        if not _check_type(maxsize, int):
-            raise ValueError("maxsize is missing.")
-        if not _check_type(top, int):
-            raise TypeError("top is not of type int.")
+    def __new__(cls, size=None, items=None, dtype=int):
+        if not _check_type(size, int):
+            raise ValueError("size is missing.")
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: If size is removed we don't need this check too.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 6]
```diff
@@ -81,47 +82,35 @@ def peek(self):
         return None
 
 class ArrayStack(Stack):
+    __slots__ = ['size', 'items', 'dtype']
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: `size` can be removed from here too.

File: pydatastructs/miscellaneous_data_structures/stack.py (hunk 4 above)
Comment (czgdp1807) [reply pydatastructs/miscellaneous_data_structures/stack.py]: We can do, `DynamicOneDimensionalArray(dtype, 0)`.

**Comment by czgdp1807** [comment]
I think we need only two data members, `items` and `dtype`.
`ArrayStack.__new__` will change to, 
```
class ArrayStack(...):
    def __new__(cls, items=None, dtype=int):
        ...
```

**Comment by Saptashrungi** [comment]
We can have only dtype instead of dtype,0 in  if items is None:

-            items = OneDimensionalArray(dtype, maxsize)
-        if not _check_type(items, OneDimensionalArray):
+            items = DynamicOneDimensionalArray(dtype, size)

**Comment by czgdp1807** [comment]
> items = DynamicOneDimensionalArray(dtype, size)

Agreed. In fact, it should be `DyamicOneDimensionalArray(dtype, 0)`.

**Comment by Saptashrungi** [comment]
Looking at the constructor of DODA I am getting confused. Is it necessary
to give the second argument after dtype? It's *arg , so will it work if no
argument other than dtype is given?

**Comment by czgdp1807** [comment]
Currently the `size` doesn't defaults to `0`, so you have to give 0 as the second argument.
```python
>>> from pydatastructs import DynamicOneDimensionalArray as DODA
>>> arr = DODA(int)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/czgdp1807/codezonediitj/pydatastructs/pydatastructs/linear_data_structures/arrays.py", line 207, in __new__
    obj = super().__new__(cls, dtype, *args, **kwargs)
  File "/home/czgdp1807/codezonediitj/pydatastructs/pydatastructs/linear_data_structures/arrays.py", line 70, in __new__
    raise ValueError("1D array cannot be created due to incorrect"
ValueError: 1D array cannot be created due to incorrect information.
```

**Comment by czgdp1807** [comment]
We can change the behaviour after some discussion but in a different PR.

**Comment by Saptashrungi** [comment]
Can we change the code of DODA, having it's default size 0? Just asking I'm
general.

**Comment by czgdp1807** [comment]
> We can change the behaviour after some discussion but in a different PR.

As said above you can do that, but in a different PR(from a different branch).

**Comment by Saptashrungi** [comment]
okay, how is my performance till now? More efforts needed, where am I
lacking?i wish to have a feedback from you.

**Comment by czgdp1807** [comment]
Thanks for the contributions first of all. It's a great experience to work with you. Hoping that you will keep communicating and contributing in this way in future too. :-)

**Comment by Saptashrungi** [comment]
Sure.

**Comment by czgdp1807** [comment]
Please update the documentation as well for the new API.

**Comment by Saptashrungi** [comment]
By documentation you means the comments in the code right?

**Comment by czgdp1807** [comment]
Yes, the comments in the code. I have made some edits in my previous comments which should be visible if you open github.

**Comment by Saptashrungi** [comment]
Will see soon.

**Comment by czgdp1807** [comment]
Please make sure that your code passes the recently added code quality tests. See the logs [here](https://travis-ci.org/codezonediitj/pydatastructs/builds/630561359?utm_source=github_status&utm_medium=notification). There are few trailing white spaces in your code.

**Comment by czgdp1807** [comment]
To test your code quality locally, update the master branch by pulling the latest changes and merge into `doda` branch of this PR.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 7]
```diff
@@ -39,15 +33,15 @@ class Stack(object):
...
     >>> s.push(1)
     >>> s.push(2)
     >>> s.push(3)
     >>> str(s)
-    '[1, 2, 3, None, None]'
+    '[1, 2, 3]'
     >>> s.pop()
     3
-
+
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: Trailing white space. Delete this line and tests will pass.

**Comment by czgdp1807** [comment]
Thanks. Merging.

**Comment by czgdp1807** [comment]
If you want you can add your name to https://github.com/codezonediitj/pydatastructs/blob/master/AUTHORS by making a separate PR.

**Pull request merged**